windows: Use GetFinalPathNameByHandle for ResolveSymbolicLink

kevpar opened the pull request:
This change splits the definition of pkg/cri/os.ResolveSymbolicLink by
platform (windows/!windows), and switches to an alternate implementation
for Windows. This aims to fix the issue described in #5405.

The previous implementation which just called filepath.EvalSymlinks has
historically had issues on Windows. One of these issues we were able to
fix in Go, but EvalSymlinks's behavior is not well specified on
Windows, and there could easily be more issues in the future, so it
seems prudent to move to a separate implementation for Windows.

The new implementation uses the Windows GetFinalPathNameByHandle API,
which takes a handle to an open file or directory and some flags, and
returns the "real" name for the object. See comments in the code for
details on the implementation.

I have tested this change with a variety of mounts and everything seems
to work as expected. Functions that make incorrect assumptions on what a
Windows path can look like may have some trouble with the \\\\?\\ path
syntax. For instance EvalSymlinks fails when given a \\\\?\\UNC\\ path. For
this reason, the resolvePath implementation modifies the returned path
to translate to the more common form (\\\\?\\UNC\\server\\share ->
\\\\server\\share).

Signed-off-by: Kevin Parsons <kevpar@microsoft.com>

AkihiroSuda commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,124 @@
+// +build windows
```
You don't need this line, because the file name has `_windows.go`

kevpar replied:
Good catch

kevpar commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+// ResolveSymbolicLink will follow any symbolic links
+func (RealOS) ResolveSymbolicLink(path string) (string, error) {
```
I added a few test cases. I'll look at what else would be good to put in there. The case that caused the issue I'm trying to fix here is pretty hard to write a good test for, since it requires creating a virtual disk and mounting it. I might be able to run something through PowerShell to do it, though.

ambarve commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,120 @@
+		// ERROR_PATH_NOT_FOUND is returned from the VOLUME_NAME_DOS query if the path is a
+		// volume GUID path to a disk, and the disk does not have an assigned drive letter.
+		// In this case, just get the volume GUID path instead.
+		rPath, err = getFinalPathNameByHandle(h, cFILE_NAME_OPENED|cVOLUME_NAME_GUID)
```
What is the difference between the volume GUID path that we used in the `openPath` and the volume GUID path that is returned here?

kevpar replied:
The string passed to `openPath` could be any sort of path string, it's not necessarily a volume GUID path.

ambarve replied:
So if the `resolvePath` is called with a path like `\\?\Volume{GUID}` then the first call to `getFinalPathNameByHandle` would fail with `ERROR_PATH_NOT_FOUND` and the second call would return the same path again?

kevpar replied:
That's correct, but that will hold true for any path resolution function. If you call it with an already fully resolved path, it should return the same thing as the input.

It might be helpful to consider the case of a symlink `lnk -> \\?\Volume{abcd}\`. In that case calling `resolvePath("lnk")` should resolve the path and return `\\?\Volume{abcd}\`.

ambarve replied:
Ah it makes sense now. I think it would be helpful to slightly change that comment to make it say `ERROR_PATH_NOT_FOUND is returned from the VOLUME_NAME_DOS query if the _final_ path is a volume GUID path to a disk`.

kevpar replied:
Thanks, that's good feedback. :) I'm actually currently working on rewriting the comments here to be much more useful. I also ended up needing to switch the order so we do VOLUME_NAME_GUID first, due to another issue in `EvalSymlinks` (and wanting to ensure the output of `resolvePath` can be passed to `EvalSymlinks` safely).

ambarve commented:
LGTM!

kevpar commented:
I've fixed up some logic, expanded the comments to explain better why we are doing it this way, and added some unit tests. However, I'm having trouble getting some of the more interesting tests to work currently, as they require creating/mounting VHDs. I initially used the Hyper-V PowerShell cmdlets, but those are not present on the CI machines, and my attempt to install them was not successful.

I'm looking into if there's another way I can do the VHD setup that is needed. Otherwise I'll probably need to take out some of the test cases, or find a way to put them behind a feature flag.

The implementation of `resolvePath` and intent behind what we are testing should still be suitable for review. @AkihiroSuda if you can take a look again that would be appreciated. :)

kevpar commented:
Okay! Finally got the tests to run properly. Everything passes now except for `pull-containerd-node-e2e`, which looks unrelated.

The only issue I see is that to properly implement VHD formatting I had to detect the Windows OS version we are running on, which brings in the `github.com/Microsoft/hcsshim/test` module as a new dependency. I'm not sure if pulling this in to the main containerd module is desired, but this test doesn't look like it would make sense under `integration/client`, either.

I could remove the version detection given the CI only runs on WS2019, but that's not ideal since someone might need to run it locally, and we don't want to prevent the CI machines from being upgraded.

I could also move the `github.com/Microsoft/hcsshim/test/functional/manifest` package out of the test module, but that will require a new hcsshim release to update to, and that might complicate backporting this to v1.4.

mainred commented:
Thanks for your effort, @kevpar . 
Thanks, @AkihiroSuda for sharing so many valuable inputs in this PR, could you please take another look?

cc @estesp

kzys commented on pkg/os/os_windows_test.go:
```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() {
+		windows.CloseHandle(windows.Handle(vhdHandle))
+	})
```
Can we use t.Cleanup? Seems moby is still using Go 1.13. #5431

kevpar replied:
I think that change won't cause issue, as it's only changing the Go version used for the `Binaries` job, and these tests only run during the `Windows Integration` job.

kzys replied:
I see. Thanks!

dims commented on go.mod:
```diff
@@ -5,6 +5,7 @@ go 1.16
 require (
 	github.com/Microsoft/go-winio v0.4.17
 	github.com/Microsoft/hcsshim v0.8.16
+	github.com/Microsoft/hcsshim/test v0.0.0-20210227013316-43a75bb4edd3
```
Can we please avoid `github.com/Microsoft/hcsshim/test` in the main go.mod?

kzys replied:
Do we have a good place for that? We have https://github.com/containerd/containerd/tree/master/integration/client but the tests in the PR are not in the integration/ folder.

dims commented on vendor/modules.txt:
```diff
@@ -41,6 +41,9 @@ github.com/Microsoft/hcsshim/internal/winapi
 github.com/Microsoft/hcsshim/pkg/ociwclayer
+# github.com/Microsoft/hcsshim/test v0.0.0-20210227013316-43a75bb4edd3
+## explicit
+github.com/Microsoft/hcsshim/test/functional/manifest
```
ugh. we will need to avoid `github.com/Microsoft/hcsshim/test/functional/manifest` as well

kevpar commented on go.sum:
```diff
@@ -53,6 +53,7 @@ github.com/Microsoft/hcsshim v0.8.15/go.mod h1:x38A4YbHbdxJtc0sF6oIz+RG0npwSCAvn
 github.com/Microsoft/hcsshim v0.8.16 h1:8/auA4LFIZFTGrqfKhGBSXwM6/4X1fHa/xniyEHu8ac=
 github.com/Microsoft/hcsshim v0.8.16/go.mod h1:o5/SZqmR7x9JNKsW3pu+nqHm0MF8vbA+VxGOoXdC600=
 github.com/Microsoft/hcsshim/test v0.0.0-20201218223536-d3e5debf77da/go.mod h1:5hlzMzRKMLyo42nCZ9oml8AdTlq/0cvIaBv6tK1RehU=
+github.com/Microsoft/hcsshim/test v0.0.0-20210227013316-43a75bb4edd3 h1:4FA+QBaydEHlwxg0lMN3rhwoDaQy6LKhVWR4qvq4BuA=
```
I'm not sure why this line didn't go away when I ran `go mod tidy`, but don't think it really matters.

cpuguy83 commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,162 @@
+)
+
+// getFinalPathNameByHandle facilitates calling the Windows API GetFinalPathNameByHandle
+// with the given handle and flags. It transparently takes care of creating a buffer of the
```
I imagine this will get called frequently.
Would it make sense to use a sync.Pool with a fixed size buffer and only create one internally if the buffer is not big enough?

cpuguy83 replied:
This would, in theory, save us an extra call to the Windows API as well as frequent allocations.

kevpar replied:
I like this idea. I've changed it to get a buffer from a pool (with default buffer size chosen to try and fit most path strings), resize if necessary, and return the final buffer to the pool.

kevpar replied:
This makes it so we only call `GetFinalPathNameByHandle` once for each query in the tests.

cpuguy83 commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,176 @@
+			b = make([]uint16, n)
+			continue
+		}
+		return windows.UTF16ToString(b), nil
```
`b[:n]`?

kevpar replied:
`UTF16ToString` will scan the slice looking for a null terminator already, so I don't think it's needed. We could change this to `string(utf16.Decode(b[:n-1]))`, though.

kevpar replied:
I changed this over, since it saves enumerating through `b` to find the null terminator, when we already know where it is.

kevpar replied:
And figured out it's `b[:n]` rather than `b[:b-1]` like I had above. :)

cpuguy83 commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,176 @@
+		// If the buffer wasn't large enough, n will be the total size needed (including null terminator).
+		// Resize and try again.
+		if n > uint32(len(b)) {
+			b = make([]uint16, n)
```
Should we return the old `b` to the pool first?

kevpar replied:
I thought maybe it would be better to just return the final (resized) buffer to the pool, so it would be picked up if the next call needs a similarly large buffer. I could add a `pool.Put(b)` here though, so that we also return the original buffer. I don't have a strong opinion on which behavior would be better. Thoughts?

cpuguy83 replied:
Probably fine like you have it.

cpuguy83 commented on pkg/os/os_windows.go:
```diff
@@ -0,0 +1,179 @@
+	New: func() interface{} {
+		// Size of buffer chosen somewhat arbitrarily to accommodate a large number of path strings.
+		// MAX_PATH (260) + size of volume GUID prefix (49) + null terminator = 310.
+		return make([]uint16, 310)
```
Linter wants this to be a pointer

kevpar replied:
Changed the pool around to hold `*[]uint16`.

cpuguy83 reviewed (approved):
#notawindowsexpert LGTM

kevpar commented:
@estesp looks like this is properly signed off. Could you please help merge?

ibabou commented:
@kevpar Is there a plan to get this cherry-picked into 1.4.x as well ? Or the only path will be into 1.5.1 ?
We currently facing issues with our csi driver on 1.4.4, so wants to see if we can get this backported.

kevpar commented:
> @kevpar Is there a plan to get this cherry-picked into 1.4.x as well ? Or the only path will be into 1.5.1 ?
> We currently facing issues with our csi driver on 1.4.4, so wants to see if we can get this backported.

@ibabou yes, I have a PR out to backport to `containerd/cri` for 1.4 here: https://github.com/containerd/cri/pull/1636

ibabou commented:
> > @kevpar Is there a plan to get this cherry-picked into 1.4.x as well ? Or the only path will be into 1.5.1 ?
> > We currently facing issues with our csi driver on 1.4.4, so wants to see if we can get this backported.
> 
> @ibabou yes, I have a PR out to backport to `containerd/cri` for 1.4 here: [containerd/cri#1636](https://github.com/containerd/cri/pull/1636)

That's awesome, thanks Kevin!

ibabou commented:
@kevpar I don't see the change in release/1.5. Is there a pending thing related to it ? Seems both 1.5.1 and 1.5.2 got published without it ?

kevpar commented:
> @kevpar I don't see the change in release/1.5. Is there a pending thing related to it ? Seems both 1.5.1 and 1.5.2 got published without it ?

It looks like the backport PR (#5454) for 1.5 accidentally went into `master` instead of `release/1.5`. I'll send a new PR to get this in. :)

kevpar commented:
> > @kevpar I don't see the change in release/1.5. Is there a pending thing related to it ? Seems both 1.5.1 and 1.5.2 got published without it ?
> 
> It looks like the backport PR (#5454) for 1.5 accidentally went into `master` instead of `release/1.5`. I'll send a new PR to get this in. :)

PR: #5537

ibabou commented:
> > > @kevpar I don't see the change in release/1.5. Is there a pending thing related to it ? Seems both 1.5.1 and 1.5.2 got published without it ?
> > 
> > 
> > It looks like the backport PR (#5454) for 1.5 accidentally went into `master` instead of `release/1.5`. I'll send a new PR to get this in. :)
> 
> PR: #5537

Yeah, I noticed too that the PR of the cherry-pick did seem to not have gone in release/1.5 :). Thanks a lot Kevin!
